pkg/solana/txm/txm.go

```diff
@@ -303,65 +295,224 @@ func (txm *Txm) sendWithRetry(ctx context.Context, msg pendingTx) (solanaGo.Tran
-					}
-				}(shouldBump, bumpCount, currentTx)
+	// Serialize the transaction message for signing
+	txMsg, marshalErr := baseTx.Message.MarshalBinary()
+	if marshalErr != nil {
+		return solanaGo.Transaction{}, fmt.Errorf("error in MarshalBinary: %w", marshalErr)
+	}
+
+	// Sign the transaction message using the key
+	sigBytes, signErr := txm.ks.Sign(ctx, key, txMsg)
+	if signErr != nil {
+		return solanaGo.Transaction{}, fmt.Errorf("error in Sign: %w", signErr)
+	}
+
+	// Append the signature to the transaction
+	var finalSig [64]byte
+	copy(finalSig[:], sigBytes)
+	baseTx.Signatures = append(baseTx.Signatures, finalSig)
+
+	return baseTx, nil
+}
+
+// getFee calculates the compute unit price for a transaction based on the number of retries.
+func (txm *Txm) getFee(count int) fees.ComputeUnitPrice {
+	// base compute unit price should only be calculated once
+	// prevent underlying base changing when bumping (could occur with RPC based estimation)
+	fee := fees.CalculateFee(
+		txm.cfg.ComputeUnitPriceDefault(),
+		txm.cfg.ComputeUnitPriceMax(),
+		txm.cfg.ComputeUnitPriceMin(),
+		uint(count), //nolint:gosec // reasonable number of bumps should never cause overflow
+	)
+	return fees.ComputeUnitPrice(fee)
+}
+
+// bumpFeeAndUpdateBlockhash determines whether the transaction fee should be bumped
+// and whether the blockhash needs to be updated based on the configured criteria.
+func (txm *Txm) bumpFeeAndUpdateBlockhash(ctx context.Context, txID string, bumpTime time.Time) (shouldBump bool, needBlockhashUpdate bool, newBumpTime time.Time, err error) {
+	// helper function to get current height
+	getCurrentHeight := func(ctx context.Context) (uint64, error) {
+		client, errClient := txm.client.Get()
+		if errClient != nil {
+			return 0, errClient
+		}
+
+		currHeight, errSlot := client.SlotHeight(ctx)
+		if errSlot != nil {
+			return 0, errSlot
+		}
+
+		return currHeight, nil
+	}
+
+	newBumpTime = bumpTime
+	currHeight, err := getCurrentHeight(ctx)
+	if err != nil {
+		return false, false, bumpTime, fmt.Errorf("failed to get current height: %w", err)
+	}
+
+	// Check if fee bumping is enabled and decide based on the criteria
+	if txm.cfg.FeeBumpPeriod() > 0 {
+		switch txm.cfg.FeeBumpCriteria() {
+		case FeeBumpCriteriaFixedIntervals:
+			// Bump the fee at fixed intervals. We may want to bump without blockhash being expired.
+			if time.Since(bumpTime) >= txm.cfg.FeeBumpPeriod() {
+				shouldBump = true
+				newBumpTime = time.Now()
 			}
 
-			// exponential increase in wait time, capped at 250ms
-			deltaT *= 2
-			if deltaT > MaxRetryTimeMs {
-				deltaT = MaxRetryTimeMs
+			// in case blockhash has expired, we need to update it.
+			if txm.txs.IsBlockhashExpired(txID, currHeight) {
+				needBlockhashUpdate = true
 			}
-			tick = time.After(time.Duration(deltaT) * time.Millisecond)
+		case FeeBumpCriteriaExpiration:
+			// Bump the fee only if the blockhash has expired
+			if txm.txs.IsBlockhashExpired(txID, currHeight) {
```

Comment:
I noticed you added expiration detection for both bumping strategies. I was thinking about it after we last talking and think this is a good idea. This might also help simplify this a bit. If we move the the expiration detection logic to a TXM level loop instead, we could actually keep that separate from gas bumping. It would be it's own separate feature that can be toggled on and off by some `TxExpirationTimeout` config (tentative name). So regardless if we have bumping or not, we could detect if txs expire and replace them with new ones with new block hashes. In that case, we could maybe leave bumping the way it is. Then we also wouldn't need an explicit `FeeBumpCriteria` config since we could implicitly set these "strategies" by either toggling fee bumping on or off and expiration detection on or off.

Reply:
I like this approach! It feels like it will simplify things.

Let’s confirm we’re on the same page before moving forward:

#### Bumping as it currently works:
- **Data Feeds** use case aiming to get transactions through as quickly as possible, even with the risk of overpaying.
- We’ll bump the existing transaction using exponential backoff.

#### Expiration feature:
- CCIP use case aiming to wait longer to avoid overpaying for transactions.
- We’ll wait until expiration to trigger `sendWithRetry` again with `baseTx` and a new `blockhash`. Although there’s no explicit bump, it may occur implicitly if `recalculated fees` are higher.

#### Interaction between them:
- If expiration and bumping are both on:
	- We’ll bump the existing transaction using exponential backoff. If expiration is reached, we’ll trigger `sendWithRetry` again with `baseTx` and a different blockhash, restarting exponential backoff bumping.
	- Reaching expiration (150 blocks) when bumping is on should be rare.

- If expiration is on and bumping is off:
	- We’ll wait until expiration to trigger `sendWithRetry` again with `baseTx` and a new `blockhash`. This won’t explicitly bump but may do so implicitly if recalculated fees are higher.

- If expiration is off and bumping is off:
	- The transaction may not go through if it expires and the fees are insufficient.

Reply:
We should be careful with `TxConfirmTimeout` though as it may overlap with `expiration`

Reply:
Notes for meeting: 

Expiration functionality changes:
- Create a loop that, every `TxExpirationTimeout`, retrieves the `slotHeight` and checks if txes in our `broadcastedMap` that are in a `broadcasted` state have expired.
- We do this by adding `lastValidBlockHeight` to the `pendingTx` when we create it. This allows us to compare the `slotHeight` to see if it has exceeded that value and decide if we need to rebuild and rebroadcast the transaction.
- To achieve this, we will need to move the blockhash assignment logic to the `sendWithRetry` function.
- If we need to rebroadcast a transaction, we should:
	- remove previous one by calling the `Remove()`
	- build the new transaction and call `sendWithRetry` with it